Add initial rough opsgenie docs (#29309)
gravitational/teleport: docs/pages/access-controls/access-request-plugins/opsgenie.mdx

---
title: Access Requests with Opsgenie
description: How to set up Teleport's Opsgenie plugin for privilege elevation approvals.
---

With Teleport's Opsgenie integration, engineers can access the infrastructure
they need to resolve alerts quickly, without longstanding admin permissions
that can become a vector for attacks.

Teleport's Opsgenie integration allows you to treat Teleport Role Access
Requests as Opsgenie alerts, notify the appropriate on-call team, and
approve or deny the requests via Teleport. You can also configure the plugin to
approve Role Access Requests automatically if the user making the request is on
the on-call team for a service affected by an alert.

This guide will explain how to set up Teleport's Access Request plugin for
Opsgenie.

> (!docs/pages/includes/plugins/diagram.mdx api="Opsgenie" action="Listen for Access Requests\nModify Access Requests" messages="Send notifications\nGet alert status" !)

Reviewer: If you look in the rendered guide, this diagram is not accurate. It shows the self-hosted plugin architecture. Let's just remove it.
```suggestion
```

## Prerequisites

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- An Opsgenie account with the ability to create API keys with the 'read' and
  'create and update' access rights.

- The Opsgenie integration is currently only available in Teleport Cloud.

- (!docs/pages/includes/tctl.mdx!)

## Step 1/5. Create services

Create an Opsgenie team named `teleport-access-request-notifications`.

We will configure the Opsgenie plugin to create an alert for the
`teleport-access-request-notifications` team when certain users
create an Access Request.

## Step 2/5. Define RBAC resources

The Teleport Opsgenie plugin works by receiving Access Request events from the
Teleport Auth Service and, based on these events, interacting with the Opsgenie
API.

### Create a requester role

To create a user first navigate to Management -> Access -> Roles

> Then select 'Create New Role' and create the requester role.

Reviewer: Let's keep the role YAML as well like you had it before so users can copy-paste.

approve Role Access Requests automatically if the user making the request is on
the on-call team for a service affected by an alert.

This guide will explain how to set up Teleport's Access Request plugin for
Opsgenie.

(!docs/pages/includes/plugins/diagram.mdx api="Opsgenie" action="Listen for Access Requests\nModify Access Requests" messages="Send notifications\nGet alert status" !)

## Prerequisites

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- An Opsgenie account with the ability to create API keys with the 'read' and
  'create and update' access rights.

- The Opsgenie integration is currently only available in Teleport Cloud.

- (!docs/pages/includes/tctl.mdx!)

## Step 1/5. Create services

Create an Opsgenie team named `teleport-access-request-notifications`.

We will configure the Opsgenie plugin to create an alert for the
`teleport-access-request-notifications` team when certain users
create an Access Request.

## Step 2/5. Define RBAC resources

The Teleport Opsgenie plugin works by receiving Access Request events from the
Teleport Auth Service and, based on these events, interacting with the Opsgenie
API.

### Create a requester role

> 

Reviewer: Since you inserted a screenshot for creating a user, should we do the same for creating a role?

```
kind: role
version: v5
metadata:
  name: requester
spec:
  allow:
    request:
      roles: ['editor']
      thresholds:
        - approve: 1
          deny: 1
      annotations:
        opsgenie_notify_services: ['teleport-access-request-notifications']
```

> The 'opsgenie_notify_services' annotation specifies the service the alert will be be created for.

Reviewer: Also, what is the proper terminology? Is it "service" or "schedule"?
```suggestion
The `opsgenie_notify_services` annotation specifies the service the alert will be created for.
```

### Create a user who will request access

Create a user called `myuser` who has the `requester` role. Later in this
guide, you will create an Access Request as this user to test the Opsgenie
plugin:

To create a user first navigate to Management -> Access -> Users

![Add user one](../../../img/enterprise/plugins/opsgenie/add-user-one.png)

Then select 'Create New User' and create a user with the requester role.

![Add user two](../../../img/enterprise/plugins/opsgenie/add-user-two.png)

## Step 3/5. Set up an Opsgenie API key

Generate an API key that the Opsgenie plugin will use to create and modify
alerts as well as list users, services, and on-call policies.

In your Opsgenie dashboard, go to **SETTINGS → API SETTINGS** and click
**API key management**. Enable 'read' and 'create and update'.

![Create API key](../../../img/enterprise/plugins/opsgenie/opsgenie-api-key.png)

## Step 4/5. Configure the Opsgenie plugin

At this point, you have generated credentials that the Opsgenie plugin will use
to connect to the Opsgenie API. To configure the plugin to use this API key navigate

> to Management -> Integrations -> Enroll New Integration.
> 
> ![Integrations menu](../../../img/enterprise/plugins/opsgenie/opsgenie-integrations-menu.png)
> 
> Then select the Opsgenie integration.

Reviewer: I would just remove the extra screenshot, I don't think it adds much.
```suggestion
to Management -> Integrations -> Enroll New Integration and select the Opsgenie integration.
```

![Integrations page](../../../img/enterprise/plugins/opsgenie/opsgenie-integration-page.png)

## Step 5/5. Test the Opsgenie plugin

### Create an Access Request

As the Teleport user `myuser`, create an Access Request for the `editor` role:

(!docs/pages/includes/plugins/create-request.mdx!)

In Opsgenie, you will see a new alert containing information about the
Access Request in either the default schedule specified when enrolling the plugin,

> or to the schedules specified by `opsgenie_notify_services` annotation in the requester's role.

Reviewer: ```suggestion
or in the schedules specified by `opsgenie_notify_services` annotation in the requester's role.
```
